Sturdy uis xstate (#78)
eggheadio/eggheadio-course-notes: construct-sturdy-uis-with-xstate/notes/03-invoke-an-xstate-service-when-entering-a-state.md

> # Invoke an Xstate Service when Enterine a state

Reviewer: ```suggestion
# Invoke an Xstate Service when Entering a state
```

- [Video link](https://egghead.io/lessons/react-invoke-an-xstate-service-when-entering-a-state)
- [Code Link](https://github.com/isaacplmann/sturdy-uis/tree/lesson2-end)

- We can reuse the `fetchMachine` in either the same component or different components.

>   - If we plan on reusing this `machine` it is good to plan ahead and decided what logic should be passed in as an argument and what logic should be delegated inside the machine.

Reviewer: ```suggestion
  - If we plan on reusing this `machine` it is good to plan ahead and decide what logic should be passed in as an argument and what logic should be delegated to inside the machine.
```

- Xstate can [invoke promises](https://xstate.js.org/docs/guides/communication.html#invoking-promises) as-is.
  - Can `resolve()` which will take `onDone` transition or it can `reject()` which will take `onError` transition.
    - `onError` data is available on the event's `data` property.
    ⚠️ If `onError` transition is missing and the `Promise` is rejected, the error will be ignored.

📝 **Note:** _If the state where the invoked promise is active is exited before the promise settles, the result of the promise is discarded._

↪️ Xstate can [invoke callbacks](https://xstate.js.org/docs/guides/communication.html#invoking-callbacks)
  - Defined as streams of events sent to the parent machine which are modeled via a callback handler. The callback handler is a function that takes two arguments:
    - `callback` - called with the event to be sent.
    - `onReceive` - called with the listern that listens to events from the parent.

🔮 Xstate can [invoke observables](https://xstate.js.org/docs/guides/communication.html#invoking-observables)
  - Defined as streams of values emmited over time.
  - Uni-directional, it can only send events to the parent machine, not receive them.

> 🚜 When xstate [invoke machines](https://xstate.js.org/docs/guides/communication.html#invoking-machines), they communicate hierarchicaclly, and invoked machines can communicate `Parent-to-child` or `Child-to-parent`

Reviewer: ```suggestion
🚜 When Xstate [invokes machines](https://xstate.js.org/docs/guides/communication.html#invoking-machines), they communicate hierarchically, and invoked machines can communicate `Parent-to-child` or `Child-to-parent`
```

>   - 🔎 Child machines can be [invoked with `context`](https://xstate.js.org/docs/guides/communication.html#invoking-machines) that is derived from the parent's machine context.

Reviewer: ```suggestion
  - 🔎 Child machines can be [invoked with `context`](https://xstate.js.org/docs/guides/communication.html#invoking-machines) that is derived from the parent machine's context.
```

```js
export const fetchMachine = Machine({
	...

	pending: {
		invoke: {
			src: 'fetchData', 👈// can define a service directly
            onDone: { target: 'successful', actions: ['setResults'] },
			onError: { target: 'failed', actions: ['setMessage'] }
		}       👆👆 // onDone and onError are both transitions
	}                // they are envents autogenerated by xState
	...

})
```

By delegating this logic to inside the machine itself, `fetchData` is no longer an action, it's a [service](https://xstate.js.org/docs/guides/communication.html).

Because we now `invoke` a `fetchData` as a service, the values stored inside `event` is now represented by a `data` property:

```js
actions: {
  setResults: assign((ctx, event: any) => ({
    results: event.data, 👈 // update from result to data

>   }));

Reviewer: ```suggestion
  })),
```

  setMessage: assign((ctx, event: any) => ({
    message: event.data, 👈 // update from message to data

>   }));

Reviewer: ```suggestion
  })),
```
